### PR TITLE
Added command history

### PR DESCRIPTION
**Pull request by AndreasArvidsson**
setting:
```json
"cursorless.commandHistory": true
```

Monthly log file: `cursorlessCommandHistory_2023-12.jsonl`
```
{"date":"2023-12-10","cursorlessVersion":"0.28.0-40a6fee4","command":{"version":6,"action":{"name":"setSelection","target":{"type":"primitive","modifiers":[{"type":"containingScope","scopeType":{"type":"line"}}]}},"usePrePhraseSnapshot":true}}
```

Voice command: `"Cursorless analyze history"` opens a new untitled document:
```
[2023-12]

Total commands: 24

Actions (7):
  setSelection: 18
  clearAndSetSelection: 1
  remove: 1
  wrapWithPairedDelimiter: 1
  getText: 1
  replace: 1
  copyToClipboard: 1

Modifiers (2):
  containingScope: 20
  relativeScope: 2

Scope types (4):
  line: 18
  surroundingPair: 2
  comment: 1
  document: 1

```

## Checklist

- [x] I have added [tests](https://www.cursorless.org/docs/contributing/test-case-recorder/)
- [x] I have updated the [docs](https://github.com/cursorless-dev/cursorless/tree/main/docs) and [cheatsheet](https://github.com/cursorless-dev/cursorless/tree/main/cursorless-talon/src/cheatsheet)
- [-] I have not broken the cheatsheet

### CODE REVIEW

File: packages/cursorless-engine/src/runCommand.ts [hunk 1]
```diff
@@ -35,6 +40,7 @@ export async function runCommand(
   languageDefinitions: LanguageDefinitions,
   rangeUpdater: RangeUpdater,
   commandRunnerDecorators: CommandRunnerDecorator[],
+  commandHistory: CommandHistory,
```
Comment (pokey) [inline packages/cursorless-engine/src/runCommand.ts:43]: Why can't this be a command decorator? (See the commandDecorators arg in the previous line (line 42))

**Review by pokey — COMMENTED** [review]
Haven't looked in depth just quick comment

File: packages/cursorless-vscode/src/ide/vscode/VscodeCommandHistory.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+  private readonly cursorlessVersion: string;
+  private disposable: Disposable;
+  private active: boolean = false;
+
+  constructor(ide: VscodeIDE, fileSystem: FileSystem) {
+    this.evaluateSetting = this.evaluateSetting.bind(this);
+    this.cursorlessVersion = ide.cursorlessVersion;
+    this.dirPath = path.join(fileSystem.cursorlessDir, dirName);
+
+    mkdirSync(this.dirPath, { recursive: true });
```
Comment (bra1nDump) [inline packages/cursorless-vscode/src/ide/vscode/VscodeCommandHistory.ts:31]: I would move the mkdir to the append method for these reasons:
- Ensures the directory is created right before writing to it. For example with the current mechanism if I delete the directory mid session, the append will throw I think
- No need to wait for directory to be created - currently blocking activate path

File: packages/cursorless-vscode/src/ide/vscode/VscodeCommandHistory.ts [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+      if (event.affectsConfiguration(settingFullName)) {
+        this.evaluateSetting();
+      }
+    });
+  }
+
+  async append(command: CommandComplete): Promise<void> {
+    if (!this.active) {
+      return;
+    }
```
Comment (bra1nDump) [inline packages/cursorless-vscode/src/ide/vscode/VscodeCommandHistory.ts:45]: We can use this to bail out instead of having DisabledCommandHistory. Seems cleaner to have a single path.

The current code will not handle enabling command history in the existing session - because we already initialized to the DisabledCommandHistory which has no workspace config watcher

Comment (AndreasArvidsson) [reply packages/cursorless-vscode/src/ide/vscode/VscodeCommandHistory.ts:45]: Yes it does. Disabled command history is only used in testing. `vscode.workspace.onDidChangeConfiguration` watches for setting changes.

File: packages/cursorless-vscode/src/ide/vscode/VscodeCommandHistory.ts (hunk 2 above)
Comment (AndreasArvidsson) [reply packages/cursorless-vscode/src/ide/vscode/VscodeCommandHistory.ts:31]: The director is owned by Cursorless and the uses should not remove it.

File: packages/cursorless-engine/src/runCommand.ts (hunk 1 above)
Comment (AndreasArvidsson) [reply packages/cursorless-engine/src/runCommand.ts]: No reason I just wanted a stricter typing. Probably unnecessary and I have re implemented it as a command runner decorator instead.

**Comment by AndreasArvidsson** [comment]
@pokey Updated and ready ;)

File: packages/cursorless-vscode/package.json [hunk 4]
```diff
@@ -279,6 +279,11 @@
           "default": true,
           "description": "Whether to show decorations on vscode start."
         },
+        "cursorless.commandHistory": {
```
Comment (pokey) [inline packages/cursorless-vscode/package.json:282]: Shall we use the `order` field to push this one up a little bit? Or wait until we're more confident in it?

File: packages/cursorless-vscode/src/CommandHistory.ts [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+  wrapCommandRunner(
+    readableHatMap: ReadOnlyHatMap,
+    runner: CommandRunner,
+  ): CommandRunner {
+    if (!this.isActive) {
+      return runner;
+    }
+
+    return {
+      run: async (commandComplete: CommandComplete) => {
```
Comment (pokey) [inline packages/cursorless-vscode/src/CommandHistory.ts:54]: Should we capture the errors? Might be useful to know how often people are running into inference issues, etc. I think just the error name would suffice. Worth quick look at how `TestCaseRecorder` wraps command in case helpful https://github.com/cursorless-dev/cursorless/blob/40a6fee4c4eec9b37cfb1ee7b4bf5ff78df59ea2/packages/cursorless-engine/src/testCaseRecorder/TestCaseRecorder.ts#L458-L483

File: packages/cursorless-vscode/src/CommandHistory.ts [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+    const historyItem: CommandHistoryItem = {
+      date: getDayDate(date),
+      cursorlessVersion: this.cursorlessVersion,
+      command: sanitizeCommand(command),
+    };
+    const data = JSON.stringify(historyItem) + "\n";
+
+    await fs.mkdir(this.dirPath, { recursive: true });
+    await fs.appendFile(file, data, "utf8");
```
Comment (pokey) [inline packages/cursorless-vscode/src/CommandHistory.ts:75]: Does this open file in exclusive mode? If not, would be worth doing so

File: packages/cursorless-vscode/src/CommandHistory.ts [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+
+  private async appendToLog(command: CommandComplete): Promise<void> {
+    const date = new Date();
+    const fileName = `${filePrefix}_${getMonthDate(date)}.jsonl`;
+    const file = path.join(this.dirPath, fileName);
+
+    const historyItem: CommandHistoryItem = {
+      date: getDayDate(date),
+      cursorlessVersion: this.cursorlessVersion,
+      command: sanitizeCommand(command),
```
Comment (pokey) [inline packages/cursorless-vscode/src/CommandHistory.ts:70]: Probably not worth changing, but `sanitizeCommand` and `sanitizeAction` would prob be cleaner using immer, like we do with canonicalization. You'd basically just change those functions to mutate the commands directly, calling them eg `sanitizeCommandInPlace`, then just use `produce` here like we do in https://github.com/cursorless-dev/cursorless/blob/40a6fee4c4eec9b37cfb1ee7b4bf5ff78df59ea2/packages/cursorless-engine/src/core/commandVersionUpgrades/canonicalizeAndValidateCommand.ts#L40-L46

File: packages/cursorless-vscode/src/CommandHistory.ts [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+function sanitizeCommand(command: CommandComplete): CommandComplete {
+  const { spokenForm, action, ...rest } = command;
+  return {
+    ...rest,
+    action: sanitizeAction(action),
+  };
+}
+
+function sanitizeAction(action: ActionDescriptor): ActionDescriptor {
+  switch (action.name) {
```
Comment (pokey) [inline packages/cursorless-vscode/src/CommandHistory.ts:99]: I know it's a bit annoying, but I would be tempted to force us to enumerate all actions here. Otherwise we may inadvertently start logging sensitive information in the future

File: packages/cursorless-vscode/src/CommandHistory.ts [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+// Remove spoken form and sanitize action
+function sanitizeCommand(command: CommandComplete): CommandComplete {
+  const { spokenForm, action, ...rest } = command;
+  return {
+    ...rest,
+    action: sanitizeAction(action),
+  };
+}
+
+function sanitizeAction(action: ActionDescriptor): ActionDescriptor {
```
Comment (pokey) [inline packages/cursorless-vscode/src/CommandHistory.ts:98]: Do we want to normalize the letter in decorated marks? Not sure how much you can glean from that, but feels like the right thing to do if people are to be comfortable sharing these things more widely. I believe we have some utilities for recursively descending commands so shouldn't be too hard I don't think

If we do, I'd argue for leaving punctuation chars, but changing all non-punctuation to just `x`

File: packages/cursorless-vscode/src/CommandHistory.ts [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+import * as fs from "node:fs/promises";
+import * as path from "node:path";
+import * as vscode from "vscode";
+
+const filePrefix = "cursorlessCommandHistory";
+const settingSection = "cursorless";
+const settingName = "commandHistory";
+const settingFullName = `${settingSection}.${settingName}`;
+
+export class CommandHistory implements CommandRunnerDecorator {
```
Comment (pokey) [inline packages/cursorless-vscode/src/CommandHistory.ts:22]: jsdoc

**Review by pokey — COMMENTED** [review]
Looks good! Left a few comments. Also

- [x] I think it would be nice to have at least one simple sunny-day test. You should be able to steal most of what you need from https://github.com/cursorless-dev/cursorless/blob/main/packages/cursorless-vscode-e2e/src/suite/testCaseRecorder.vscode.test.ts

File: packages/cursorless-vscode/package.json (hunk 4 above)
Comment (AndreasArvidsson) [reply packages/cursorless-vscode/package.json:282]: Sounds reasonable. What would be a good order do you think?

File: packages/cursorless-vscode/src/CommandHistory.ts (hunk 9 above)
Comment (AndreasArvidsson) [reply packages/cursorless-vscode/src/CommandHistory.ts:98]: I was thinking about that, but I also think that which letters you use is an interesting data point?

File: packages/cursorless-vscode/src/CommandHistory.ts (hunk 7 above)
Comment (AndreasArvidsson) [reply packages/cursorless-vscode/src/CommandHistory.ts:70]: I don't know. Personally I find that pattern quite hard to read.

File: packages/cursorless-vscode/src/CommandHistory.ts (hunk 6 above)
Comment (pokey) [reply packages/cursorless-vscode/src/CommandHistory.ts:75]: @josharian what did you mean when you said "open the file in exclusive mode"? From what I understand, exclusive just means fail if the file already exists, which I don't think is what we want here

File: packages/cursorless-vscode/src/CommandHistory.ts (hunk 8 above)
Comment (AndreasArvidsson) [reply packages/cursorless-vscode/src/CommandHistory.ts:99]: updated

File: packages/cursorless-vscode/src/CommandHistory.ts (hunk 7 above)
Comment (pokey) [reply packages/cursorless-vscode/src/CommandHistory.ts:70]: I find deeply nested spreads hard to read. What don't you like about the immer pattern?

Just to be clear, let's not block this PR on this discussion, but I'm interested to have it anyway

File: packages/cursorless-vscode/src/CommandHistory.ts (hunk 9 above)
Comment (pokey) [reply packages/cursorless-vscode/src/CommandHistory.ts:98]: I could go either way. @josharian any thoughts?

File: packages/cursorless-vscode/src/CommandHistory.ts (hunk 10 above)
Comment (AndreasArvidsson) [reply packages/cursorless-vscode/src/CommandHistory.ts:25]: Not quite sure what to write in this case but I added a small comment

File: packages/cursorless-vscode/src/CommandHistory.ts (hunk 8 above)
Comment (pokey) [reply packages/cursorless-vscode/src/CommandHistory.ts:102]: Yeah I think generally we should try to avoid them if we can, but in this case I think it's justified

**Comment by AndreasArvidsson** [comment]
> Looks good! Left a few comments. Also
> 
>     * [ ]  I think it would be nice to have at least one simple sunny-day test. You should be able to steal most of what you need from https://github.com/cursorless-dev/cursorless/blob/main/packages/cursorless-vscode-e2e/src/suite/testCaseRecorder.vscode.test.ts

I can't find the mechanism for mocking the file system paths.

**Comment by pokey** [comment]
> I can't find the mechanism for mocking the file system paths.

The cursorless dir is already automatically set to a temp dir https://github.com/cursorless-dev/cursorless/blob/40a6fee4c4eec9b37cfb1ee7b4bf5ff78df59ea2/packages/cursorless-vscode/src/extension.ts#L175-L177 and then you can access it like this https://github.com/cursorless-dev/cursorless/blob/40a6fee4c4eec9b37cfb1ee7b4bf5ff78df59ea2/packages/cursorless-vscode-e2e/src/suite/scopeProvider/runCustomSpokenFormScopeInfoTest.ts#L12-L13

File: packages/cursorless-vscode/src/CommandHistory.ts (hunk 6 above)
Comment (josharian) [reply packages/cursorless-vscode/src/CommandHistory.ts:78]: sorry about the confusion. I always mess this up. ideally we'd use something like a lockfile to coordinate access, e.g. https://www.npmjs.com/package/proper-lockfile (found via google search, no personal experience). it's either that or hope, which isn't necessarily a terrible approach here.

File: packages/cursorless-vscode/src/CommandHistory.ts (hunk 9 above)
Comment (josharian) [reply packages/cursorless-vscode/src/CommandHistory.ts:101]: i don't feel strongly.

File: packages/cursorless-vscode/src/CommandHistory.ts (hunk 7 above)
Comment (AndreasArvidsson) [reply packages/cursorless-vscode/src/CommandHistory.ts:73]: That is just so the code should be as compact as possible. You can definitely make it more verbose if you prefer that pattern. 

I think that immer adds a layer of abstraction that is quite esoteric. Instead of just returning or assigning a value all of the sudden you have a callback with an argument and no return value. Instead of constructing a new object without a field you should now do `delete draft.spokenForm`. In my mind immer solves a problem that we don't really have here and goes against the kiss principle.

File: packages/cursorless-vscode/package.json (hunk 4 above)
Comment (AndreasArvidsson) [reply packages/cursorless-vscode/package.json:282]: That setting seems completely broken to me. I have shapes and hat stability at the top
image

File: packages/cursorless-vscode/src/CommandHistory.ts (hunk 5 above)
Comment (AndreasArvidsson) [reply packages/cursorless-vscode/src/CommandHistory.ts:57]: But we are not logging the result? Should we log the command itself and then optionally log errors?

**Comment by AndreasArvidsson** [comment]
> > I can't find the mechanism for mocking the file system paths.
> 
> The cursorless dir is already automatically set to a temp dir
> 
> https://github.com/cursorless-dev/cursorless/blob/40a6fee4c4eec9b37cfb1ee7b4bf5ff78df59ea2/packages/cursorless-vscode/src/extension.ts#L175-L177
> and then you can access it like this
> 
> https://github.com/cursorless-dev/cursorless/blob/40a6fee4c4eec9b37cfb1ee7b4bf5ff78df59ea2/packages/cursorless-vscode-e2e/src/suite/scopeProvider/runCustomSpokenFormScopeInfoTest.ts#L12-L13

Ok but I do wonder how I can make command history be active only for this single test. I don't want it logging every single test in our recorded suite.

File: packages/cursorless-vscode/package.json (hunk 4 above)
Comment (pokey) [reply packages/cursorless-vscode/package.json:282]: Try saying "cursorless settings"

**Comment by pokey** [comment]
> Ok but I do wonder how I can make command history be active only for this single test. I don't want it logging every single test in our recorded suite.

You can mock the setting. See eg https://github.com/cursorless-dev/cursorless/blob/8f88564b880c246ba226981218722d300dabb965/packages/cursorless-vscode-e2e/src/suite/keyboard/basic.vscode.test.ts#L139-L155

You may need to trigger a setting refresh. Worth testing also that nothing is written if the setting is off. That seems important from privacy perspective

File: packages/cursorless-vscode/src/CommandHistory.ts (hunk 5 above)
Comment (pokey) [reply packages/cursorless-vscode/src/CommandHistory.ts:57]: Not sure I follow. You could just add an optional error field to the top-level payload

File: packages/cursorless-vscode/src/CommandHistory.ts (hunk 6 above)
Comment (pokey) [reply packages/cursorless-vscode/src/CommandHistory.ts:78]: Yeah honestly just "hope" might be fine, since I'm not sure I see how you'd end up with multiple windows writing at once

Comment (josharian) [reply packages/cursorless-vscode/src/CommandHistory.ts:78]: yeah. talon and the command file are both serving as de facto external locking mechanisms here. if we end up with corrupt logs, we can reconsider.

Comment (pokey) [reply packages/cursorless-vscode/src/CommandHistory.ts:78]: And even by keyboard, you won't be typing into multiple windows

Comment (josharian) [reply packages/cursorless-vscode/src/CommandHistory.ts:78]: one certainly hopes. people do weird things. (and win weird prizes when they do.)

File: packages/cursorless-vscode/src/CommandHistory.ts (hunk 5 above)
Comment (AndreasArvidsson) [reply packages/cursorless-vscode/src/CommandHistory.ts:57]: I just have my mind set on logging before executing the command. I guess we could add it afterwards.
Updated

File: packages/cursorless-vscode/package.json (hunk 4 above)
Comment (AndreasArvidsson) [reply packages/cursorless-vscode/package.json:282]: Probably first time I have used at command. Now we got the desired order, but that also means that unless you get to the settings in a particular way the order doses matter. Weird.

File: packages/cursorless-vscode/src/CommandHistory.ts (hunk 6 above)
Comment (AndreasArvidsson) [reply packages/cursorless-vscode/src/CommandHistory.ts:89]: Then I guess we're happy?

File: packages/cursorless-vscode/src/CommandHistory.ts (hunk 9 above)
Comment (AndreasArvidsson) [reply packages/cursorless-vscode/src/CommandHistory.ts:112]: Then are we okay with leaving it as it is? I think it would be interesting to see which letters a user prefer. Or if there are letters/symbols that you don't use at all.

File: packages/cursorless-vscode/src/CommandHistory.ts (hunk 6 above)
Comment (pokey) [reply packages/cursorless-vscode/src/CommandHistory.ts:89]: Yeah I think let's leave it and keep an eye out for corrupt json. Not seeing any better solution that's worth the effort

File: packages/cursorless-vscode/src/CommandHistory.ts (hunk 9 above)
Comment (pokey) [reply packages/cursorless-vscode/src/CommandHistory.ts:112]: Still makes me nervous. Writing user's document text to file seems sketchy, even if it's a single character at a time. And I'm not sure what we'll glean from knowing which char it is. I would be inclined to let non-alphanumeric pass through untouched, and letters map to `x` and numbers map to `0`. I guess we could ask on slack to see how comfortable people are?

Comment (AndreasArvidsson) [reply packages/cursorless-vscode/src/CommandHistory.ts:112]: I don't count this as document text at all. The only thing we know is that the character exists. How many times and in which configurations we don't know. This is at maximum fingerprinting the user's language. I don't see a point to only keep non alphanumerics. We either remove everything or nothing is my vote.
Sounds reasonable to ask on slack.

File: packages/cursorless-vscode/src/CommandHistory.ts (hunk 7 above)
Comment (pokey) [reply packages/cursorless-vscode/src/CommandHistory.ts:84]: Let's not block this PR on this discussion. I did a quick draft of what this would look like with immer in https://github.com/cursorless-dev/cursorless/pull/2124; let's discuss that one at our leisure. Not urgent at all

File: packages/cursorless-vscode/src/CommandHistory.ts (hunk 9 above)
Comment (josharian) [reply packages/cursorless-vscode/src/CommandHistory.ts:112]: Given that it is opt-in and local-only, I think it’s probably OK to record the char.

Comment (pokey) [reply packages/cursorless-vscode/src/CommandHistory.ts:112]: Ok fine let's keep it

**Comment by pokey** [comment]
I believe this is just waiting on tests and docs, right? Fwiw I reactivated the "tests" checkbox; curious why you disabled it in the first place?

**Comment by AndreasArvidsson** [comment]
> I believe this is just waiting on tests and docs, right? Fwiw I reactivated the "tests" checkbox; curious why you disabled it in the first place?

Yes. Because I didn't think there was any way this could be tested.

**Comment by AndreasArvidsson** [comment]
@pokey Tests added. Not sure where to put the documentation though?

**Comment by AndreasArvidsson** [comment]
@pokey Added a small docs page

File: packages/cursorless-vscode/src/extension.ts [hunk 11]
```diff
@@ -99,6 +100,8 @@ export async function activate(
     fileSystem,
   );
 
+  addCommandRunnerDecorator(new CommandHistory(context, vscodeApi, fileSystem));
```
Comment (pokey) [inline packages/cursorless-vscode/src/extension.ts:103]: How come this is VSCode-only? I know it's easier to put things in VSCode directory, but I think we should strive to put as much as we can in the engine

File: cursorless-talon/src/cursorless.py [hunk 12]
```diff
@@ -15,3 +15,9 @@ def private_cursorless_show_settings_in_ide():
 
     def private_cursorless_show_sidebar():
         """Show Cursorless-specific settings in ide"""
+
+    def private_cursorless_analyze_command_history():
```
Comment (pokey) [inline cursorless-talon/src/cursorless.py:19]: I think I'd call this `private_cursorless_show_statistics`. Also, not sure this needs an action, but I guess there's not real downside 🤷‍♂️

File: cursorless-talon/src/cursorless.talon [hunk 13]
```diff
@@ -43,3 +43,6 @@ tag: user.cursorless
 
 bar {user.cursorless_homophone}:
     user.private_cursorless_show_sidebar()
+
+{user.cursorless_homophone} analyze history:
```
Comment (pokey) [inline cursorless-talon/src/cursorless.talon:47]: maybe
```suggestion
{user.cursorless_homophone} show stats:
```

File: packages/cursorless-engine/src/index.ts [hunk 14]
```diff
@@ -6,3 +6,7 @@ export * from "./core/StoredTargets";
 export * from "./typings/TreeSitter";
 export * from "./cursorlessEngine";
 export * from "./api/CursorlessEngineApi";
+export * from "./CommandRunner";
+export * from "./util/getPartialTargetDescriptors";
+export * from "./util/getPrimitiveTargets";
+export * from "./util/getScopeType";
```
Comment (pokey) [inline packages/cursorless-engine/src/index.ts:12]: A bit of a smell to have to export all these; I really think as much of command history should be engine-side

File: packages/cursorless-vscode/src/CommandHistoryAnalyzer.ts [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+  }
+
+  getMonths() {
+    const months = Object.values(this.months);
+    months.sort((a, b) => a.month.localeCompare(b.month));
+    return months;
+  }
+}
+
+async function analyzeFilesAndReturnMonths(dir: string): Promise<Month[]> {
```
Comment (pokey) [inline packages/cursorless-vscode/src/CommandHistoryAnalyzer.ts:111]: Why are all your abstractions in terms of months? Feels a bit arbitrary. The roll-up window should just be some knob we can tweak, rather than being deeply baked into the analysis code.

Shall we factor this analysis code into a separate PR so we can review that separately without blocking this one? I'd love to get the history part merged in

**Review by pokey — COMMENTED** [review]
Made some tweaks and left a few comments; almost there

File: packages/cursorless-vscode/src/extension.ts (hunk 11 above)
Comment (AndreasArvidsson) [reply packages/cursorless-vscode/src/extension.ts:103]: I had it in the engine originally, but since we changed it to a runner decorator there is no hard engine dependency anymore. It also needs a vscode config so the other option would be to migrate this to our own ide.configuration object?

File: cursorless-talon/src/cursorless.py (hunk 12 above)
Comment (AndreasArvidsson) [reply cursorless-talon/src/cursorless.py:19]: The setting is called command history and I wanted the command to reflect that they belong together? Without a spoken command people will not use this I'm afraid

File: packages/cursorless-vscode/src/CommandHistoryAnalyzer.ts (hunk 15 above)
Comment (AndreasArvidsson) [reply packages/cursorless-vscode/src/CommandHistoryAnalyzer.ts:111]: Just an arbitrary choice that I felt suited it. Easy to change if we want to.
There's no reason why we can't change the analysis code later?

File: packages/cursorless-engine/src/index.ts (hunk 14 above)
Comment (AndreasArvidsson) [reply packages/cursorless-engine/src/index.ts:12]: I guess we could move the Cursorless version and the command history setting to the engine?

File: cursorless-talon/src/cursorless.talon (hunk 13 above)
Comment (pokey) [reply cursorless-talon/src/cursorless.talon:47]: tho we might want to go object-verb? Eg "stats show" or "history stats"?

File: packages/cursorless-engine/src/index.ts (hunk 14 above)
Comment (pokey) [reply packages/cursorless-engine/src/index.ts:12]: yeah I'd be tempted to just move it all to the engine. No reason to arbitrarily make it harder for other ides in the future

File: cursorless-talon/src/cursorless.py (hunk 12 above)
Comment (AndreasArvidsson) [reply cursorless-talon/src/cursorless.py:19]: Oh you mean we could just execute this rpc command directly? Yeah you're right we should probably just remove this action

File: packages/cursorless-vscode/src/CommandHistoryAnalyzer.ts (hunk 15 above)
Comment (pokey) [reply packages/cursorless-vscode/src/CommandHistoryAnalyzer.ts:111]: I guess it's very self-contained, but I just don't like landing code I haven't reviewed, and I'd love to get this thing in

File: cursorless-talon/src/cursorless.py (hunk 12 above)
Comment (AndreasArvidsson) [reply cursorless-talon/src/cursorless.py]: Unnecessary private action is now removed.

File: packages/cursorless-vscode/src/extension.ts (hunk 11 above)
Comment (AndreasArvidsson) [reply packages/cursorless-vscode/src/extension.ts:103]: We can do that

File: packages/cursorless-vscode/src/CommandHistoryAnalyzer.ts (hunk 15 above)
Comment (AndreasArvidsson) [reply packages/cursorless-vscode/src/CommandHistoryAnalyzer.ts:111]: I've gonna have to take the time to move everything to the engine so you have time to look at this if you want to.

File: packages/cursorless-engine/src/index.ts (hunk 14 above)
Comment (AndreasArvidsson) [reply packages/cursorless-engine/src/index.ts:12]: Agreed

Comment (pokey) [reply packages/cursorless-engine/src/index.ts:12]: Hmm. Let's put it there for now

File: cursorless-talon/src/cursorless.py (hunk 12 above)
Comment (josharian) [reply cursorless-talon/src/cursorless.py]: > Idk I find "show statistics" more intuitive. cc/ @josharian you have a leaning?af

My leaning is not to have a spoken command yet, while we mess around to find out what is useful, and then hope that once a handful of us have played a bit, it'll be clearer (a) what we want to show and thus (b) what a good spoken form for that is.

File: packages/cursorless-vscode/src/CommandHistoryAnalyzer.ts (hunk 15 above)
Comment (pokey) [reply packages/cursorless-vscode/src/CommandHistoryAnalyzer.ts:111]: Atomic PRs

Comment (pokey) [reply packages/cursorless-vscode/src/CommandHistoryAnalyzer.ts:111]: Atomic PRs

**Comment by pokey** [comment]
I think it would be good hygiene to include a command id and phrase id in each log entry.

Also, it would be good to test that our sanitisation actually works. I could see that breaking without us noticing 😅

**Comment by AndreasArvidsson** [comment]
> I think it would be good hygiene to include a command id and phrase id in each log entry.
> 
> Also, it would be good to test that our sanitisation actually works. I could see that breaking without us noticing 😅

We don't really have the command id at this point in the pipeline, but I guess we could just use the constant with the knowledge that it will be correct?
What do you mean with a phrase id?

True sounds reasonable to test.

**Comment by AndreasArvidsson** [comment]
> > We don't really have the command id at this point in the pipeline, but I guess we could just use the constant with the knowledge that it will be correct?
> 
> No I just mean generate one on the fly

Okay how would that id look and what would be the purpose? Why not just use the actual command id?

> > What do you mean with a phrase id?
> 
> Just generate an id, then use command server api to see if the phrase has changed, like we do for hat map, and if not reuse previous phrase id otherwise generate new one. Make sense?

Yes. If it's the same phrase use the same id otherwise generate new one. Using this we can group commands in the phrase.

**Comment by pokey** [comment]
> Okay how would that id look and what would be the purpose? Why not just use the actual command id?

Just a uuid I guess? I thought you said we didn't have access to command id. The purpose is just hygiene. If we do transformations on the files, it's nice for each command to have an id so we can be sure it's the same. Just a data science best practice

> Yes. If it's the same phrase use the same id otherwise generate new one. Using this we can group commands in the phrase.

👍

**Comment by AndreasArvidsson** [comment]
> > Okay how would that id look and what would be the purpose? Why not just use the actual command id?
> 
> Just a uuid I guess? I thought you said we didn't have access to command id. The purpose is just hygiene. If we do transformations on the files, it's nice for each command to have an id so we can be sure it's the same. Just a data science best practice

I don't really see how a uuid that gets generated for each command helps with anything? Or do you mean that we should just have a static one like a version number?
We don't have access to the command in the command runner itself, but we only have a single command to run Cursorless actions and that is it exported constant.
https://github.com/cursorless-dev/cursorless/blob/6718e84eb033e880980929098d4b21d2ff2613f2/packages/common/src/cursorlessCommandIds.ts#L1-L2

**Comment by pokey** [comment]
> I don't really see how a uuid that gets generated for each command helps with anything?

It's just good practice. These IDs are very useful when doing postprocessing. But we can see if others feel strongly we should leave them out cc/ @phillco @josharian

File: packages/vscode-common/src/TestHelpers.ts [hunk 16]
```diff
@@ -46,6 +46,7 @@ export interface TestHelpers {
   runIntegrationTests(): Promise<void>;
 
   cursorlessTalonStateJsonPath: string;
```
Comment (AndreasArvidsson) [inline packages/vscode-common/src/TestHelpers.ts:48]: Should we return the file system instance instead of keep adding paths here?

**Comment by AndreasArvidsson** [comment]
Made a few changes
* moved command history to the engine
* added Cursorless version string to the ide
* added phrase id to command history entry
* removed spoken command for analyzing command history

File: cursorless-talon/src/cursorless.py (hunk 12 above)
Comment (AndreasArvidsson) [reply cursorless-talon/src/cursorless.py]: Sounds good

**Comment by pokey** [comment]
update from meet-up:

- add id for each payload

**Comment by AndreasArvidsson** [comment]
id added

File: packages/vscode-common/src/TestHelpers.ts (hunk 16 above)
Comment (pokey) [reply packages/vscode-common/src/TestHelpers.ts:48]: Let's leave for now and revisit if it gets out of hand

**Comment by pokey** [comment]
Removed analyzer; created https://github.com/cursorless-dev/cursorless/issues/2161 to track re-adding. Feel free to open PR with that patch once this one is in; lmk if you want help creating the patch @AndreasArvidsson

**Review by pokey — APPROVED** [review]
Ok I finished those last two steps, and tried installing locally to make sure it works as expected. Looking good! Feel free to merge if you're happy with my changes
